### PR TITLE
Added reranker.py

### PR DESCRIPTION
**Pull request by iujinasena**
Please review, @deltaaruna .

### CODE REVIEW

File: effectzgpt/backend/app/engine/__init__.py
Comment (deltaaruna) [inline effectzgpt/backend/app/engine/__init__.py:1]: Pls add these things
1) In app.engine.node_postprocessors, define a method to return the reranker(We have only cohere reranknig implemented, so check if the cohere api key is available, if so, return the cohere reranker from reranker.py)
2) Check of the reranking is enabled. If enabled, get the reranker. Something like the following.
if os.getenv("IS_RERANKER_ENABLED", "True").lower() == "true":
        node_postprocessors.append(get_reranker())

Pls refer to sentence window retrieval implementation for more info.

**Review by deltaaruna — CHANGES_REQUESTED** [review]
Nice job.
Please look at my comment and implement it.

File: effectzgpt/backend/app/engine/constants.py [hunk 1]
```diff
@@ -4,4 +4,5 @@
 DEFAULT_WINDOW_META_DATA_KEY = "window"
 # Default original text metadata for sentence window retrieval
 ORIGINAL_TEXT_METADATA_KEY = "original_text"
-
+# Top k for retrieval
+DEFAULT_TOP_K = 3
```
Comment (nilaan-effectz) [inline effectzgpt/backend/app/engine/constants.py]: use DEFAULT_COHERE_TOP_K  and refactor usages

File: effectzgpt/backend/app/engine/reranker.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import os
+from app.engine.constants import DEFAULT_TOP_K
+from llama_index.postprocessor.cohere_rerank import CohereRerank
+
+def get_cohere_reranker():
+    api_key = os.getenv("COHERE_API_KEY")
+    top_k = int(os.getenv("TOP_K", DEFAULT_TOP_K))
```
Comment (nilaan-effectz) [inline effectzgpt/backend/app/engine/reranker.py]: TOP_K is already in use , use COHERE_TOP_K 
add these env variable to .env file

File: effectzgpt/backend/app/engine/__init__.py [hunk 3]
```diff
@@ -26,7 +26,7 @@ def get_chat_engine(filters=None):
         Do not respond to any queries that are not related to Jordanian government services.
  """
     
-    cohere_rerank = CohereRerank(api_key=cohere_api_key, top_n=10)
+    cohere_rerank = get_cohere_reranker()
```
Comment (nilaan-effectz) [inline effectzgpt/backend/app/engine/__init__.py]: remove this line 
in line 42  'node_postprocessors.append(cohere_rerank)'
use if condition 
`
if os.getenv("IS_RERANKER_ENABLED").lower() == "true" :
    cohere_rerank = get_cohere_reranker()
   node_postprocessors.append(cohere_rerank)
`
make sure you added this env varible to .env file (default set it to false )

File: effectzgpt/backend/app/engine/__init__.py
Comment (iujinasena) [reply effectzgpt/backend/app/engine/__init__.py:1]: Requested changes have been implemented. Please review, @deltaaruna .

**Review by nilaan-effectz — CHANGES_REQUESTED** [review]

**Comment by deltaaruna** [comment]
Looks good. Please resolve the conflicts. Then this can be merged. Thanks @iujinasena

**Review by nilaan-effectz — APPROVED** [review]

**Pull request merged**